test/test_model.py

```diff
@@ -732,6 +734,133 @@ def assertBackendCalls(self, expected, *, reset=True):
+            container = containers[name]
+            self.assertIsInstance(container, ops.model.Container)
+            self.assertEqual(container.name, name)
+            self.assertIsInstance(container.pebble, ops.pebble.Client)
+        with self.assertRaises(KeyError):
+            containers['c3']
+
+        with self.assertRaises(RuntimeError):
+            other_unit = self.model.get_unit('other')
+            other_unit.containers
+
+    def test_unit_get_container(self):
+        unit = self.model.unit
+        for name in ['c1', 'c2']:
+            container = unit.get_container(name)
+            self.assertIsInstance(container, ops.model.Container)
+            self.assertEqual(container.name, name)
+            self.assertIsInstance(container.pebble, ops.pebble.Client)
+        with self.assertRaises(ops.model.ModelError):
+            unit.get_container('c3')
+
+        with self.assertRaises(RuntimeError):
+            other_unit = self.model.get_unit('other')
+            other_unit.get_container('foo')
+
+
+class TestContainerPebble(unittest.TestCase):
+    def setUp(self):
+        meta = ops.charm.CharmMeta.from_yaml("""
+name: k8s-charm
+containers:
+  c1:
+    k: v
+""")
+        backend = ops.model._ModelBackend('myapp/0')
+        with unittest.mock.patch('ops.pebble.Client', MockPebbleClient):
+            self.model = ops.model.Model(meta, backend)
+        self.container = self.model.unit.containers['c1']
+        self.pebble = self.container.pebble
+
+    def test_socket_path(self):
+        self.assertEqual(self.pebble.socket_path, '/charm/containers/c1/pebble/.pebble.socket')
+
+    def test_autostart(self):
+        self.container.autostart()
+        self.assertEqual(self.pebble.requests, [('autostart',)])
+
+    def test_start(self):
+        self.container.start('foo')
+        self.container.start('foo', 'bar')
+        self.assertEqual(self.pebble.requests, [
+            ('start', ('foo',)),
+            ('start', ('foo', 'bar')),
+        ])
+
+    def test_stop(self):
+        self.container.stop('foo')
+        self.container.stop('foo', 'bar')
+        self.assertEqual(self.pebble.requests, [
+            ('stop', ('foo',)),
+            ('stop', ('foo', 'bar')),
+        ])
+
+    def test_add_layer(self):
+        self.container.add_layer('summary: str\n')
+        self.container.add_layer({'summary': 'dict'})
+        self.container.add_layer(ops.pebble.Layer('summary: Layer'))
+        self.assertEqual(self.pebble.requests, [
+            ('add_layer', 'summary: str\n'),
+            ('add_layer', 'summary: dict\n'),
+            ('add_layer', 'summary: Layer\n'),
+        ])
+
+    def test_get_layer(self):
+        self.pebble.responses.append('summary: foo')
+        layer = self.container.get_layer()
+        self.assertEqual(self.pebble.requests, [('get_layer',)])
+        self.assertIsInstance(layer, ops.pebble.Layer)
+        self.assertEqual(layer.summary, 'foo')
+
+
+class MockPebbleClient:
```

Comment:
import mock

is part of the stdlib, and helps avoid some of the boilerplate here, but also tends to have a little bit too much magic. Mock with spec_set tends to avoid some of the obvious pitfalls (it makes sure the signatures of the methods are correct, rather than just accepting anything)
Something to consider, at least.

Reply:
I've given this a look, and I guess I don't love `Mock`, as it requires you to learn a whole new DSL instead of just writing a few lines of clear Python. We're not using it elsewhere in the project, so there's not a precedent. Happy to switch in another PR if we really decide to go all in on `Mock`.